Extend `gradlogpdf` to MixtureModels

rmsrosa opened the pull request:
This extends `gradlogpdf` to MixtureModels, both univariate and multivariate, at least for those whose components have `gradlogpdf` implemented.

I haven't implemented the inplace and the component-wise methods yet, but this should be a good start.

I should say I am having trouble with the docs, thought. I have added docstrings and added the methods to the docs src, but they are not showing up. I am not sure what to do.

sethaxen commented:
Relates #1788

rmsrosa commented:
> Relates #1788

Oh, I missed that. Sorry.

I thought about using `pweights`, but I was afraid it would allocate. I should benchmark both to make sure.

devmotion commented on src/common.jl:
```diff
@@ -277,6 +277,46 @@ end
+
+See also: [`pdf`](@ref).
+"""
+@inline function gradlogpdf(
+    d::Distribution{ArrayLikeVariate{N}}, x::AbstractArray{<:Real,M}
+) where {N,M}
```
My feeling is that the implicit broadcasting behaviour of `logpdf` is a bit unfortunate and that `logpdf` on multiple variates should be deprecated and instead users should be pointed to `eachvariate` (+ `map` or `broadcast`). Therefore I'm a bit hesitant to add this generalization for `gradlogpdf`.

rmsrosa replied:
Oh, ok. I was just trying to follow along, but now I understand the TO DO comment to deprecate it. I can remove this part (probably a week from now, after a trip).

devmotion commented on src/mixtures/mixturemodel.jl:
```diff
@@ -359,15 +367,22 @@ function _mixlogpdf!(r::AbstractArray, d::AbstractMixtureModel, x)
 end
 
+function _mixgradlogpdf1(d::AbstractMixtureModel, x)
+    glp = insupport(d, x) ? sum(pi * pdf(d.components[i], x) .* gradlogpdf(d.components[i], x) for (i, pi) in enumerate(probs(d)) if (!iszero(pi) && !iszero(pdf(d.components[i], x)))) / pdf(d, x) : zero(x)
```
I expect that in many cases this definition introduces a type instability (since generally the return types of the two branches will be different).

rmsrosa replied:
I see. Makes sense. I will have another look when I get the chance. Thanks for the quick feedback.

rmsrosa replied:
I believe the following would be an alternative to avoid type instability, right?

```
function Distributions.gradlogpdf(d::UnivariateMixture, x::Real)
          pdfx = pdf(d, x)
          glp = !iszero(pdfx) ? sum(pi * pdf(d.components[i], x) .* gradlogpdf(d.components[i], x) for (i, pi) in enumerate(probs(d)) if (!iszero(pi) && !iszero(pdf(d.components[i], x)))) / pdfx : pdfx
           return glp
       end
```

rmsrosa commented:
But I feel this new version is not optimized. It essentially computes the PDF twice. I should unroll the loop.

rmsrosa commented on src/mixtures/mixturemodel.jl:
```diff
@@ -367,15 +367,31 @@ function _mixlogpdf!(r::AbstractArray, d::AbstractMixtureModel, x)
+function gradlogpdf(d::UnivariateMixture, x::Real)
+    cp = components(d)
+    pr = probs(d)
+    pdfx1 = pdf(cp[1], x)
+    pdfx = pr[1] * pdfx1
+    _glp = pdfx * gradlogpdf(cp[1], x)
+    glp = (!iszero(pr[1])) && (!iszero(pdfx)) ? _glp : zero(_glp)
+    @inbounds for i in Iterators.drop(eachindex(pr, cp), 1)
```
I am not sure how to avoid duplicating the operations. I have to have `glp` out of the loop and initiated with correct type. I will think more about it and about how to use `iterate`. I actually like `Iterators.drop`, but if there is a way to avoid duplicating the code, that's better.

devmotion replied:
What I meant is: You retrieve the first elements (and the state) with `iterate(pr)` and `iterate(cp)` outside of the loop (and then you already have the elements and don't need to call `cp[1]` etc. multiple times), and then in the for loop (maybe rather a while loop then) you would call `iterate(pr, pr_state)` etc. to retrieve the next element and the new state (and again you don't need to call `pr[i]` etc. multiple times).

rmsrosa replied:
Ok. I see. Is there any performance advantage with that or is it more a matter of taste? I will give it a try anyway. Thanks.

devmotion replied:
I don't expect a performance difference (the indexing shouldn't matter compared with the actual computational cost) but IMO there are a few other advantages: It does not assume 1-based indexing, it avoids `Iterators.drop` etc, and the iteration/indices of both arrays are kept completely separate.

rmsrosa replied:
Oh, of course, I should have used `first()` instead of `[1]`. But, yeah, `iterate` each one does not assume they have a common indexing. Good.

rmsrosa replied:
This is the best I came up with.

I don't like much `while true` loops, but that is what `sum()` essentially does (it boils down to https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L53) and it avoids some extra calsl to `iterate()`.

rmsrosa commented:
But I am not sure about type stability. Even `pdf` itself is not always type stable, e.g.
```julia
julia> @code_warntype pdf(MixtureModel([Normal(1, 2), Beta(2, 3)], [6/10, 4/10]), 1.0)
MethodInstance for Distributions.pdf(::MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, ::Float64)
  from pdf(d::UnivariateMixture, x::Real) @ Distributions ~/.julia/packages/Distributions/UaWBm/src/mixtures/mixturemodel.jl:362
Arguments
  #self#::Core.Const(Distributions.pdf)
  d::MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}
  x::Float64
Body::Any
1 ─ %1 = Distributions._mixpdf1(d, x)::Any
└──      return %1
```

with
```julia
@code_warntype Distributions._mixpdf1(MixtureModel([Normal(1, 2), Beta(2, 3)], [6/10, 4/10]), 1.0)
MethodInstance for Distributions._mixpdf1(::MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, ::Float64)
  from _mixpdf1(d::AbstractMixtureModel, x) @ Distributions ~/.julia/packages/Distributions/UaWBm/src/mixtures/mixturemodel.jl:286
Arguments
  #self#::Core.Const(Distributions._mixpdf1)
  d::MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}
  x::Float64
Locals
  #607::Distributions.var"#607#609"
  #606::Distributions.var"#606#608"{MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, Float64}
  p::Vector{Float64}
Body::Any
1 ─       (p = Distributions.probs(d))
│   %2  = Distributions.:(var"#606#608")::Core.Const(Distributions.var"#606#608")
│   %3  = Core.typeof(d)::Core.Const(MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}})
│   %4  = Core.typeof(x)::Core.Const(Float64)
│   %5  = Core.apply_type(%2, %3, %4)::Core.Const(Distributions.var"#606#608"{MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, Float64})
│         (#606 = %new(%5, d, x))
│   %7  = #606::Distributions.var"#606#608"{MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, Float64}
│         (#607 = %new(Distributions.:(var"#607#609")))
│   %9  = #607::Core.Const(Distributions.var"#607#609"())
│   %10 = Distributions.enumerate(p)::Base.Iterators.Enumerate{Vector{Float64}}
│   %11 = Base.Filter(%9, %10)::Base.Iterators.Filter{Distributions.var"#607#609", Base.Iterators.Enumerate{Vector{Float64}}}
│   %12 = Base.Generator(%7, %11)::Base.Generator{Base.Iterators.Filter{Distributions.var"#607#609", Base.Iterators.Enumerate{Vector{Float64}}}, Distributions.var"#606#608"{MixtureModel{Univariate, Continuous, Distribution{Univariate, Continuous}, Categorical{Float64, Vector{Float64}}}, Float64}}
│   %13 = Distributions.sum(%12)::Any
└──       return %13
```

rmsrosa commented:
Hmm, I thought I had added another comment on benchmark, but maybe the connection broke and it didn't go through. Here it goes again.

I also implemented another version with the for loop detaching the indices of both `prior` and `components`.

```julia
function gradlogpdf(d::UnivariateMixture, x::Real)
    ps = probs(d)
    cs = components(d)
    ps1 = first(ps)
    cs1 = first(cs)
    pdfx1 = pdf(cs1, x)
    pdfx = ps1 * pdfx1
    glp = pdfx * gradlogpdf(cs1, x)
    if iszero(ps1) || iszero(pdfx)
        glp = zero(glp)
    end
    @inbounds for (psi, csi) in Iterators.drop(zip(ps, cs), 1)
        if !iszero(psi)
            pdfxi = pdf(csi, x)
            if !iszero(pdfxi)
                pipdfxi = psi * pdfxi
                pdfx += pipdfxi
                glp += pipdfxi * gradlogpdf(csi, x)
            end
        end
    end
    if !iszero(pdfx) # else glp is already zero
        glp /= pdfx
    end 
    return glp
end
```

When mixing distributions of the same type, `gradlogpdf` with the **while loop** is a tad faster, but, when  mixing different types of distributions, then `pdf` and `logpdf` are type unstable, making `gradlogpdf` also type unstable and in this case `gradlogpdf` with the **for loop** is faster. (Same thing for the multivariate versions of the functions). (I used `@btime` and `@benchmark` to make sure.) Here are some examples.

```julia
d1 = MixtureModel([Normal(1, 2), Normal(2, 3), Normal(1.5), Normal(1.0, 0.2)], [0.3, 0.2, 0.3, 0.2])

d2 = MixtureModel([Beta(1, 2), Beta(2, 3), Beta(4, 2)], [3/10, 4/10, 3/10])

d3 = MixtureModel([Normal(1, 2), Beta(2, 3), Exponential(3/2)], [3/10, 4/10, 3/10]) # type unstable

n = 10
m = 20
d4 = MixtureModel([MvNormal(rand(n), rand(n, n) |> A -> (A + A')^2) for _ in 1:m], rand(m) |> w -> w / sum(w))

d5 = MixtureModel([MvNormal([1.0, 2.0], [0.4 0.2; 0.2 0.5]), MvTDist(5., [1., 2.], [1. 0.1; 0.1 1.])], [0.4, 0.6]) # type unstable
```

We get
```julia
[ Info: d1: gradlogpdf with for loop
  75.841 ns (0 allocations: 0 bytes)
[ Info: d1: gradlogpdf with while loop
  73.626 ns (0 allocations: 0 bytes)

[ Info: d2: gradlogpdf with for loop
  295.552 ns (0 allocations: 0 bytes)
[ Info: d2: gradlogpdf with while loop
  296.826 ns (0 allocations: 0 bytes)

[ Info: d3: gradlogpdf with for loop
  686.398 ns (19 allocations: 304 bytes)
[ Info: d3: gradlogpdf with while loop
  974.368 ns (21 allocations: 368 bytes)

[ Info: d4: gradlogpdf with for loop
  33.258 μs (136 allocations: 16.62 KiB)
[ Info: d4: gradlogpdf with while loop
  33.233 μs (136 allocations: 16.62 KiB)

[ Info: d5: gradlogpdf with for loop
  3.282 μs (26 allocations: 1.42 KiB)
[ Info: d5: gradlogpdf with while loop
  3.471 μs (28 allocations: 1.53 KiB)
```

Ok, I think this is it. I will leave it with the *while loop* for your review.

devmotion reviewed (commented):
The performance difference is a bit surprising since `for` loops are rewritten using the `iterate` anyway (you can see this e.g. by inspecting the output of `@code_lowered`).

My gut feeling is also that one cannot expect too much performance wise in the type unstable case anyway. In this use-case it might be better to work with tuples (which IIRC currently is not supported).

rmsrosa commented on src/mixtures/mixturemodel.jl:
```diff
@@ -362,6 +369,41 @@ end
+            end
+        end
+    end
+    if !iszero(pdfx) # else glp is already zero
+        glp /= pdfx
+    end 
```
I wonder about that, but decided to follow the current behavior already implemented. For example:
```julia
julia> insupport(Beta(0.5, 0.5), -1)
false

julia> gradlogpdf(Beta(0.5, 0.5), -1)
0.0
```

rmsrosa replied:
Most implementations return zero outside the support:
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/beta.jl#L126
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/chi.jl#L99
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/chisq.jl#L94
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/exponential.jl#L97

and so on (Frechet, Gamma, ...), but some (Arcsin and Kumaraswamy) seem to just extend what happens at the edges of the support:
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/arcsine.jl#L97
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/kumaraswamy.jl#L91

while LogitNormal seems to be the only one that yields `NaN`:
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/laplace.jl#L101

Meanwhile, Laplace seems to be the only one to error at a singular point:
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/laplace.jl#L101

while Uniform, for example, is zero everywhere including the singular points.
https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/uniform.jl#L86

I think giving an error is a bit drastic, but yielding either zero or NaN (outside the support or at the singular points) seems reasonable to me.

rmsrosa replied:
Maybe we should make this a separate issue and hopefully standardize the behavior.
